scripts/Muon/GUI/Common/threading_manager.py

```diff
@@ -0,0 +1,282 @@
+        :param kwargs: Keyword argument lists to pass to the function (e.g. x =[1,2,3] for f(x) where x is an integer).
+        """
+        super(Worker, self).__init__()
+        self.fn = threading_decorator(fn)
+        self.kwargs = kwargs
+        self.signals = WorkerSignals()
+        # Add callback so function can update its progress
+        self.kwargs['progress_callback'] = self.signals.progress
+
+    def run(self):
+        self.signals.started.emit()
+        try:
+            result, fails = self.fn(**self.kwargs)
+            if 'progress_callback' in self.kwargs:
+                del self.kwargs['progress_callback']
+            for fail in fails:
+                # Exceptions caught whilst executing the function
+                exctype, value = type(fail), next(iter(fail.args), "")
+                self.signals.error.emit({"inputs": self.kwargs,
+                                         "exctype": exctype,
+                                         "value": value,
+                                         "traceback": traceback.format_exc()})
+        except:
+            # Catch errors not thrown by the function evaluation
+            exctype, value = sys.exc_info()[:2]  # traceback.print_exc()
+            if 'progress_callback' in self.kwargs:
+                del self.kwargs['progress_callback']
+            self.signals.error.emit({"inputs": self.kwargs,
+                                     "exctype": exctype,
+                                     "value": value,
+                                     "traceback": traceback.format_exc()})
+        else:
+            # Successful evaluation
+            if 'progress_callback' in self.kwargs:
+                del self.kwargs['progress_callback']
+            self.signals.result.emit((self.kwargs, result))
+        finally:
+            self.signals.finished.emit()
+
+    def cancel(self):
+        self.signals.cancelled.emit()
+        # must use terminate (not quit/exit) as there is no event loop for the thread
+        self.terminate()
+
+
+class WorkerManager(QtWidgets.QWidget):
+    """
+    The WorkerManager class handles multi-threading of a function, by splitting an arbitrary number of list arguments
+    equally across a specified number of threads.
+
+    Example : for a function f(x,y) where x and y are ints, the WorkerManager allows for evaluation of
+    f(x=[1,2,3,4,5],y=[1,2,3,4,5]) by splitting the lists across the given number of threads. So for two threads the
+    following evaluations might occur:
+        Thread 1 : f(1,1), f(3,3), f(5,5)
+        Thread 2 : f(2,2), f(4,4)
+
+    A list of callbacks can be passed into the constructor to respond to events (all threads finishing,
+    threads updating on progress, threads catching exceptions from the function evaluation)
+    """
+
+    finished = Signal()  # used for unit tests
+    cancelled = Signal()  # used for unit tests
+
+    def __init__(self, fn, num_threads,
```

Comment:
Yes, and default of `1` is probably sensible so that multiple threads are only created if specifically requested and the default behaviour is just to start a single "background" thread.